outgoing IP per domain

runout-at opened the pull request:
this is still missing the part for the webUI.
But i want to discuss if this is going in the right direction.

#30

runout-at committed:
add field out_ip

outgoing ip addresses separated by ';'

runout-at committed:
add OUTGOING_IP_ADDRESSES

runout-at committed:
add out_ip

for outgoing ip addresses

runout-at committed:
fix to sender domain

because this is for outgoing mail we have to lookup the senders IP.
I think this should be taken from the AUTH of the user who logged in to send this message.

runout-at committed:
fix $domain<->$sender_address_domain

last fix was in the wrong line

Udera commented:
You are planning that you can put a list of IPs? Or how do you manage ipv4 and ipv6?

runout-at commented:
The easiest solution would be to have just one additional field in the 'domains' table and put a semicolon separated list of IPs there. This way we can have mix of IPv4 and IPv6 in this field and can use it 'as is' in the transport like `interface = <; OUTGOING_IP_ADDRESSES`

Maybe the WebUI could be more complex if we like to check/sanitize the values on input or split ddresses in separate fields - but to me, one field would be enough. I wouldn't do checks on every mail in exim.
One could monitor the validity of the addresses with a cron job.

runout-at committed:
mon-mailoutip

bash script for checking if vexim outgoing ips exists in systems network config

runout-at committed:
Delete mon-mailoutip

runout-at committed:
monitoring script mon-mailoutip

checks the out-IPs from the vexim domains table if they are configured on any systems network interface

runout-at committed:
bugfix

bugfix for message text
config now even works with ansible out of the box

runout-at committed:
check for bad ip adresses

check if an ip address has valid syntax using ipv6calc

runout-at commented:
This is on my production servers for 2 weeks without issues, IPv4/v6 dual stack.

The only part missing here is the transports.
For now i add following to the original Debian transport files but i'm not sure what is the best way to integrate this in vexim:
```
.ifdef OUTGOING_IP_ADDRESSES
  interface = <; OUTGOING_IP_ADDRESSES
.endif
```
added to following files:
/etc/exim4/conf.d/transport/30_exim4-config_remote_smtp
/etc/exim4/conf.d/transport/30_exim4-config_remote_smtp_smarthost

runout-at committed:
latest version from production machine

runout-at committed:
fix: exit if no IPs returnd from mysql

runout-at committed:
fix for empty mysql lookup

Udera commented:
Let's wait for the 2.4 milestone when the new webinterface is planned. Though it seems very far away.

runout-at commented:
i think that we can have the actual changes earlier and do the webUI later.

Having different IPs per domain is a good feature. Besides from IPv6 it makes it easier getting mails through spam filters at gmail and others - eg if some domains on a server are sending bulk mails, newsletters,... it's better other mail is going out over a different IP.

Commit:
Merge branch 'master' into runout-at-outip

merge master

runout-at commented:
A rework of this PR makes sense since Debian has integrated this in the default configs now.

I did use the variable _OUTGOING_IP_ADDRESSES_. Debian uses _REMOTE_SMTP_INTERFACE_. That's the only difference.

We only really need the changes in the DB and a config setting to read the IP from the database like:

- /etc/exim4/conf.d/main/00_vexim_listmacrosdefs

```
VEXIM_OUTIP = SELECT DISTINCT out_ip FROM domains \
                           WHERE domain = '${quote_mysql:$sender_address_domain}' \
                               AND out_ip <> TRIM('')
OUTGOING_IP_DEFAULT = 2001:db8:: ; 10.0.0.1
REMOTE_SMTP_INTERFACE = ${lookup mysql{VEXIM_OUTIP}{ <; $value }{ <; OUTGOING_IP_DEFAULT }}
```

Udera commented:
How does the smtp transport know which ip to use (or is this already hidden somewhere)? You just add all of the ip addresses:

> The first interface of the correct type (IPv4 or IPv6) is used for the outgoing connection.

https://www.exim.org/exim-html-current/doc/html/spec_html/ch-the_smtp_transport.html

runout-at commented:
In Debian split config its in transport/30_exim4-config_remote_smtp

```
.ifdef REMOTE_SMTP_INTERFACE
  interface = REMOTE_SMTP_INTERFACE
.endif
```

My code from above does set the variable _REMOTE_SMTP_INTERFACE_ from the DB-query or a default value.

runout-at commented:
After reading your comment again, I'm not sure I did understand your question right. Maybe my answer was not sufficient.

If multiple IPs are given in _REMOTE_SMTP_INTERFACE_, Exim will try each until it finds one that is sufficient. Eg if the connection is IPv4 and the first addresses given are IPv6 it will skip it. At least it works like that on my production sever for several years now.
